### PR TITLE
Feature/refactor

### LINKED ISSUE #17: Rebuild Runners to be Functionally designed
Runners will be redesigned to be functionally designed,

This should allow a user to insert code at any point within the inference loop

### PR DESCRIPTION
**Pull request by b8raoult**
## Refactor `anemoi.inference`

Closes #17 , #9 

----
📚 Documentation preview 📚: https://anemoi-inference--33.org.readthedocs.build/en/33/

### CODE REVIEW

File: src/anemoi/inference/runner.py [hunk 1]
```diff
@@ -391,186 +235,121 @@ def run(
...
             # Predict next state of atmosphere
-            with torch.autocast(device_type=device, dtype=autocast):
-                y_pred = model.predict_step(input_tensor_torch)
+            with torch.autocast(device_type=self.device, dtype=self.autocast):
+                y_pred = self.model.predict_step(input_tensor_torch)
 
-            # Detach tensor and squeeze
-            output = np.squeeze(y_pred.cpu().numpy())
+            # Detach tensor and squeeze (should we detach here?)
+            output = np.squeeze(y_pred.cpu().numpy())  # shape: (values, variables)
```
Comment (HCookie) [inline src/anemoi/inference/runner.py:264]: I think detaching should be an optional behaviour

File: src/anemoi/inference/runner.py [hunk 2]
```diff
@@ -391,186 +235,121 @@ def run(
...
-            # Compute new forcing
-
-            forcing = forcing_and_constants(
-                source=input_fields[:1],
-                param=self.checkpoint.computed_forcings,
-                date=start_datetime + datetime.timedelta(hours=step),
-            )
-            forcing = np.swapaxes(forcing[np.newaxis, np.newaxis, ...], -2, -1)
-            forcing = torch.from_numpy(forcing).to(device)
+            # Update  tensor for next iteration
```
Comment (HCookie) [inline src/anemoi/inference/runner.py:272]: I think due to the different ways to prepare for the next iteration, this functionality should be abstracted out.

File: src/anemoi/inference/runner.py (hunk 1 above)
Comment (b8raoult) [reply src/anemoi/inference/runner.py:267]: Let's merge and refactor in another PR

File: src/anemoi/inference/runner.py (hunk 2 above)
Comment (b8raoult) [reply src/anemoi/inference/runner.py:278]: Let's merge and refactor in another PR

**Pull request merged**